89. Gray Code

```diff
@@ -0,0 +1,26 @@
+//Time: O (2 ^ n) Space: O(2 ^ n)
```

Comment:
regardless all comments above, since you have a clear algorithm here. However, do you think you need recursion here?

Reply:
I first solution on my mind is that I should use DFS, which is recursive in nature. I come up with the current algorithm on half way writing the DFS, and found that I do not actually need recursive function when I finished it.

Reply:
If this is not optimized code and clearly you know what it should look like, I will strongly suggest you write a clean version without recursion. And please do so in the interview. Show them your clear logic and write concise and readable code.

Reply:
First I want to verify that this comment is on Word Ladder. Looks like git hub has a problem here..
And it appears to me that I am not writing enough code that "To come up with a solution" is not that easy for me. I only come up with a bad solution, and the better ones does not come up in the first tens of minutes. And when the problem is kind of difficult, I will spend too much focus on  what should I do with the algorithm and some of the details, like n or n-1 is not carefully manipulated.
I will try to as your suggested. And I think I should have more comments..
Another question, how much time do I have to finish a question like word ladders in the interview?

Reply:
well, I believe with FB's full-time employee standard, it should take 20-25 mins, including introducing, writing code, testing, and optimization. It is ok for you to take 1h to come up with the clean solution. I think what you need to do next time is going through your code again and again, questioning your solution in any way you can think of. For example, why do I think this variable? Can this logic be simpler? Can I use less memory? Why do I have to use recursion here? This is different from your previous experience gained from school or standard test(GRE/高考), a correct answer is not what we want. A clean logic, solid reasoning and nice code are what we are looking for.

Reply:
One common mistake people may encounter is that they try to find the best solution at the beginning. That's not the usual way. The best solution usually comes from original bad solution. After you refining the solution over and over again and questioning your approach a few times, you will find the best solution. That's what we called optimization during an interview.

Reply:
Thank you. You really come to the point. I usually save the program after it is accepted on Leetcode, as long as the big O time and space is optimized. In another word, the optimizations that does not affect the big O time and space are not carefully treated. I should have been more careful with every problem. I am, actually, worried since I started to do the Leetcode, and considers the number of problems matters.
For the timing part, actually it takes loooong time to come up with a good solution. Partially because that I am not in the "Test Mode", but the time is also loonger than you expect. I cannot really come up with a solution for some problems (e.g 337. House Robber III). 
I will try to do optimization first, and then reduce the writing time. 
BTW, I have not done 高考.

Reply:
It is ok to come up with solutions slower than expected. When you think more and more, your brain will adapt to the way of discrete mathematics. Same reason, it is hard for me to develop PDE model for some physical or EE problem. Just because I am not familiar with those fields, even though I know all the essential stuff in PDE.